### ISSUE TITLE
Error joining node to cluster when it has joined another trusted cluster

### ISSUE THREAD

**Issue by ekristen** [gravitational/teleport#896]
`can not join: certificate generation error: remote-cluster has no signing keys`

`tctl auth ls` shows a CA Type for both user and host.

**Comment by ekristen** [comment]
It seems as though once I join a cluster as a trusted cluster to another one, I cannot add anymore nodes to the cluster without getting the error above.

**Comment by ekristen** [comment]
@russjones or @kontsevoy have you seen this?

**Comment by klizhentas** [comment]
@ekristen I haven't seen this, but we can try to reproduce on our side if you provide us with step-by step guide and tell us what version are you using

**Comment by ekristen** [comment]
**Version:** RC2 (mainly because all the releases since have not had any builds associated with them)

I am using dynamic tokens only to register hosts.

1. Get your first cluster up and running (I'm using dynamodb as a backend)
2. Then get your second cluster up and running, using local storage only (no HA just a single auth server)
3. Join the second cluster to the first cluster using the dynamic trusted-cluster config option (see below for trusted config file)
4. Attempt to join a second node to the second cluster, it will result in the error above.

```
kind: trusted_cluster
version: v1
metadata:
  description: "Trusted Cluster my-second-env"
  name: "my-second-env"
  namespace: "default"
spec:
  enabled: true
  roles: [ "admin" ]
  token: one-time-generated-with-trusted-cluster-role
  tunnel_addr: my-first-cluster-proxy:3024
  web_proxy_addr: my-first-cluster-proxy:3080
```

The second cluster successfully joins and I can see the primary node (that happens to run the auth server and the proxy instance for the reverse tunnel, but once you try and add any other node to the cluster it errors saying there are no keys available for signing.

**Comment by ekristen** [comment]
I just upgraded to v2.0.1 and the problem still exists. I'm doing some more debugging, I might have more information in the morning.

**Comment by ekristen** [comment]
It seems that if you create a cluster and have it join another cluster as a trusted cluster using the dynamic config option before adding a node to the cluster it will error out with no signing keys. 

However if you create your cluster, add at least one node to the cluster first then tell the auth/proxy service to join another cluster as a trusted_cluster using dynamic config everything works correctly.

**Comment by ekristen** [comment]
Unfortunately however I am now unable to join nodes to my primary cluster, I get the same error

```
    level=error msg="[Node] failed to join the cluster: primary-cluster has no signing keys" file="service/service.go:599" func="service.(*TeleportProcess).RegisterWithAuthServer.func1"
```

**Comment by kontsevoy** [comment]
@ekristen this happens because `/var/lib/teleport` (AKA "data dir") contains host certificates issued by another node (my guess). I would recommend wiping out /var/lib/teleport if a node changes cluster membership. LMK if that helps.

**Comment by kontsevoy** [comment]
@ekristen do you mind sending us YAML configs (feel free to strip/replace sensitive info in them) for the following:

* Auth server of cluster-a
* Auth sever of cluster-b
* Node of cluster-b which fails to join

Thanks!

**Comment by ekristen** [comment]
@kontsevoy most definitely. Just to be clear, I know have cluster b working , but cluster a has stopped working. I'm not sure if that's because of the upgrade to 2.0.1 or if it is because of of getting cluster b working properly.

## Original Problem 
1. I setup cluster-A, joined nodes to it, everything worked create
2. Setup cluster-B, joined it to cluster-A as trusted cluster
3. Attempted to add new node to cluster-B, fails with **cluster-B has no signing keys**

## Current Problem
1. Upgraded cluster-A to 2.0.1
2. Re-setup cluster-B with 2.0.1, joined nodes to cluster-B, everything worked
3. Joined cluster-B to cluster-A, this worked
4. Attempted to add new node to cluster-A, fails with **cluster-A has no signing keys**

### cluster-a auth config
```
teleport:
  data_dir: /var/lib/teleport
  auth_token: TOKEN
  advertise_ip: 10.X.X.X
  connection_limits:
    max_connections: 250
    max_users: 25
  log:
    output: stderr
    severity: ERROR
  storage:
    type: dynamodb
    region: us-west-2
    table_name: teleport
    access_key: XXXXXXXX
    secret_key: YYYYYYYY

proxy_service:
  enabled: no

auth_service:
  enabled: yes
  listen_addr: 0.0.0.0:3025
  cluster_name: cluster-a
  dynamic_config: true
  authentication:
    type: local
    second_factor: otp
  tokens:
    - "auth,node:TOKEN"

ssh_service:
  enabled: yes
  listen_addr: 0.0.0.0:3022
  labels:
    environment: my-env
  commands:
    - name: hostname
      command: [/bin/hostname]
      period: 1m0s
    - name: arch
      command: [/bin/uname, -p]
      period: 1h0m0s
```

### cluster-b auth
```
teleport:
  data_dir: /var/lib/teleport
  auth_token: TOKEN
  advertise_ip: 10.y.y.y
  connection_limits:
    max_connections: 250
    max_users: 25
  log:
    output: /var/log/teleport.log
    severity: INFO

proxy_service:
  enabled: yes
  listen_addr: 0.0.0.0:3023
  web_listen_addr: 0.0.0.0:3080

auth_service:
  enabled: yes
  listen_addr: 0.0.0.0:3025
  cluster_name: cluster-b
  dynamic_config: true
  authentication:
    type: local
    second_factor: otp
  tokens:
    - "auth,proxy,node:TOKEN"

ssh_service:
  enabled: yes
  listen_addr: 0.0.0.0:3022
  labels:
    environment: cluster-b
  commands:
    - name: hostname
      command: [/bin/hostname]
      period: 1m0s
    - name: arch
      command: [/bin/uname, -p]
      period: 1h0m0s
```

### cluster-a node config
this is failing to join with the error message of "has no signing keys"
```
teleport:
  data_dir: /var/lib/teleport
  auth_token: AUTH_TOKEN
  auth_servers:
    - auth0.cluster-a.private.io:3025
  advertise_ip: 10.z.z.z
  connection_limits:
    max_connections: 250
    max_users: 25
  log:
    output: /var/log/teleport.log
    severity: INFO

proxy_service:
  enabled: no

auth_service:
  enabled: no

ssh_service:
  enabled: yes
  listen_addr: 0.0.0.0:3022
  labels:
    environment: cluster-a
  commands:
    - name: hostname
      command: [/bin/hostname]
      period: 1m0s
    - name: arch
      command: [/bin/uname, -p]
      period: 1h0m0s
```

### cluster-b node config
```
teleport:
  data_dir: /var/lib/teleport
  auth_token: AUTH_TOKEN
  auth_servers:
    - auth0.cluster-b.private.io:3025
  advertise_ip: 10.y.y.y
  connection_limits:
    max_connections: 250
    max_users: 25
  log:
    output: /var/log/teleport.log
    severity: INFO

proxy_service:
  enabled: no

auth_service:
  enabled: no

ssh_service:
  enabled: yes
  listen_addr: 0.0.0.0:3022
  labels:
    environment: cluster-b
  commands:
    - name: hostname
      command: [/bin/hostname]
      period: 1m0s
    - name: arch
      command: [/bin/uname, -p]
      period: 1h0m0s
```

**Comment by ekristen** [comment]
I was about to blow away the dynamodb table and all my cluster-A `/var/lib/teleport` directories on all servers and start over, but if you'd like me to hold off to try and gather more information please let me know.

**Comment by kontsevoy** [comment]
@ekristen I realize you've done some editing, but this jumped out anyway: `auth_token` (on the node) and `tokens` on the auth do not match. Is that as a result of editing?

**Comment by ekristen** [comment]
Result of editing, my apologies.  I'm using saltstack and I've writing custom modules that actually use salt stack to generate a 3m ttl token with the appropriate roles and have it populate the auth_token field of the new node, and then start the service up.

The only hard coded token I have is for the auth servers and the reason for this is before the auth server are online there is no auth service to talk to with tctl to generate a token to join with.

**Comment by ekristen** [comment]
I've definitely seen the error if the token is no longer valid, and that is an entirely different error message than what I'm seeing. I can provide more logs if you'd like.

**Comment by ekristen** [comment]
@kontsevoy if you've got a slack server, irc room, or discord server that you idle in, I'd be more than willing to join so when you have time we can talk real-time, should that be more beneficial to you.

**Comment by kontsevoy** [comment]
@ekristen lets do this first. can you try to reproduce this from scratch using "dir" backend? I.e. on all machines use this back-end config:

```
teleport:
   storage:
       type: dynamodb
```

Then, if the problem can be reproduced, you can just zipup `/var/lib/teleport` on both auth servers and send it to us securely. Thanks!

**Comment by ekristen** [comment]
I can do that. Are you sure that backend config is correct though? Seems to be referencing dynamodb still?

**Comment by kontsevoy** [comment]
you probably read my previous message via email notification? :) because I have corrected it right away :) should be:

```
teleport:
   storage:
       type: dir
```

**Comment by ekristen** [comment]
Funny enough I was actually on the github page but for some reason your comment didn't get updated. 

I'll get this tested ASAP, most likely tomorrow morning and let you know.

**Comment by ekristen** [comment]
So far a fresh install with a single auth service seems to be working, so there might be a bug in the HA code? I'll keep testing and adding new hosts and let you know if anything changes.

**Comment by kontsevoy** [comment]
hm... could be something in the upgrade path. so this works on DynamoDB now as well?

**Comment by ekristen** [comment]
I have not tried dynamo yet. I've been just going through and testing using dir. 

Quick question, it's only the auth servers that have to talk to dynamodb?

**Comment by kontsevoy** [comment]
@ekristen correct. DynamoDB is auth server's storage back-end. nodes and proxies are more or less stateless (their state is read-only and is created once upon the initial run).

**Comment by ekristen** [comment]
Everything seems to be working with storage dir and a single auth server. I'll delete everything and try dynamodb again and let you know what happens.

**Comment by ekristen** [comment]
Well I upgraded to alpha8 and now I cannot add anymore nodes. Getting the cluster has no signing keys.

**Comment by ekristen** [comment]
There seems to be something with upgrading to a new version using dynamodb that breaks everything.

**Comment by ekristen** [comment]
```level=warning msg="[AUTH] Node \"server-001\" [11fbfa42-17b9-4cfe-a863-65e791663838] can not join: certificate generation error: my-cluster has no signing keys" file="auth/auth.go:464" func="auth.(*AuthServer).GenerateServerKeys"```

**Comment by kontsevoy** [comment]
@ekristen looking into it. Can you give us the output of `teleport version` on the auth server before and after upgrade to the alpha? Thanks.

**Comment by ekristen** [comment]
I cannot since I've already upgraded. However I was on 2.0.6 and now I am on `Teleport v2.2.0-alpha.8 git:v2.1.0-alpha.6-43-g14cf169d-dirty` 

I can tell you that I've now seen this happen multiple times across multiple versions. I am using dynamodb as a backend.  I attempted to replicate it using `dir` mode only and a single auth server and was unable to.

After that I went back to using dynamodb with mulitple auth servers however I only use 1 when registering nodes, so while the other 2 are running the auth service nothing is talking to them. 

Everything seemed great for a while, I was able to add nodes and this bug seemed to not be present anymore until I upgraded to alpha8 and I completely lost the ability to register nodes again.
